src/main/java/school/lemon/changerequest/java/introduction/hw2/Task9.java

```diff
@@ -1,7 +1,24 @@
 package school.lemon.changerequest.java.introduction.hw2;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 public class Task9 {
     public static String[] extractData(String URL) {
+        if (URL == null) {
+            return new String[]{"", "", ""};
+        }
+        Pattern p = Pattern.compile("(\\w+)://(\\w+\\-?\\w+\\.?\\w+):?(\\d+)?");
+        List<String> list = new ArrayList();
```

Comment:
No need to use loop and list here at all.
Also, `://` are specified only when specified schema, so it should be in one optional group. The same for port and `:`.
`\\w` is not good choice for host. It's better to use exclusions of some characters like `:`, `/` and spaces from all characters.